org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -573,12 +577,31 @@ protected Compiler newCompiler() {
 	CompilerOptions compilerOptions = new CompilerOptions(projectOptions);
 	compilerOptions.performMethodsFullRecovery = true;
 	compilerOptions.performStatementsRecovery = true;
-	Compiler newCompiler = new Compiler(
-		this.nameEnvironment,
-		DefaultErrorHandlingPolicies.proceedWithAllProblems(),
-		compilerOptions,
-		this,
-		ProblemFactory.getProblemFactory(Locale.getDefault()));
+
+	ICompilerFactory compilerFactory = null;
+	String compilerFactoryClassName = System.getProperty(COMPILER_FACTORY_KEY);
+	if (compilerFactoryClassName != null) {
+		try {
+			Class<? extends ICompilerFactory> compilerFactoryClass = (Class<? extends ICompilerFactory>) Class.forName(compilerFactoryClassName);
+			Constructor<? extends ICompilerFactory> constructor = compilerFactoryClass.getDeclaredConstructor();
+			compilerFactory = constructor.newInstance();
+		} catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException
+				| IllegalArgumentException | InstantiationException | InvocationTargetException e) {
+			ILog.get().error("Failed to initialize the custom compiler factory - " + compilerFactoryClassName, e); //$NON-NLS-1$
+		}
+	}
+
+	if (compilerFactory == null) {
+		compilerFactory = new DefaultCompilerFactory();
+	}
+
+	Compiler newCompiler = compilerFactory.newCompiler(
+			this.nameEnvironment,
+			DefaultErrorHandlingPolicies.proceedWithAllProblems(),
+			prepareCompilerConfiguration(compilerOptions),
```

Comment:
Since the default ECJ implementation only relies on the `compilerOptions` property from the `CompilerConfiguration`, and `prepareCompilerConfiguration()` retains the original compilerOptions when creating the CompilerConfiguration. All other properties computation in it is for the experimental flag. I believe it will not impact the default implementation.

Reply:
In the future perhaps this prepare logic could also be moved to the factory, to feed each compiler implementation those options it needs. Or you could perhaps demonstrate that `CompilerConfiguration` is indeed an abstraction suitable also for ecj, beyond that one field `compilerOptions`. But for the time being this design is acceptable. 

`prepareCompilerConfiguration()` just mustn't throw exceptions (looks fairly safe) nor cause performance issues (looks fairly fast).